### PR TITLE
AB#119 add endurance powerup

### PR DESCRIPTION
**Pull request by sir-marc**

### CODE REVIEW

File: darwin-types/src/game-objects/index.ts [hunk 1]
```diff
@@ -2,3 +2,4 @@ export * from './GameObject';
 export * from './Consumable';
 export * from './Unit';
 export * from './Attributes';
+export { default as PowerupType } from './Powerup';
```
Comment (StuckiSimon) [inline darwin-types/src/game-objects/index.ts]: Don't do default exports in the types. Stick with named exports.

File: game-backend/src/game-engine/mechanics/powerup-spawner/index.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  const unitCount = countUnits(state);
+  const powerupCount = countPowerups(state);
+
+  const powerupSpawnLimit = Math.floor(unitCount * POWERUP_TO_UNIT_RATIO);
+
+  if (powerupCount < powerupSpawnLimit) {
+    const position = generateFreePosition(state);
+    const food = spawnPowerup(position);
+
+    return {
```
Comment (StuckiSimon) [inline game-backend/src/game-engine/mechanics/powerup-spawner/index.ts]: Really small point:
What about using produce?
```
return produce(state, draft => {
  draft.objectMap[food.id] = food;
  draft.objectIds.push(food.id)
})
```

File: game-backend/src/game-engine/mechanics/powerup-spawner/index.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import { State } from '@darwin/types';
+import spawnPowerup from './spawnPowerup';
+import { countUnits } from '../../../helper/gameObjects';
+import { generateFreePosition } from '../../../helper/fields';
+import { countPowerups } from '../../../helper/powerup';
+
+const POWERUP_TO_UNIT_RATIO = 0.5;
+
+const handlePowerupSpawning = (state: State): State => {
```
Comment (StuckiSimon) [inline game-backend/src/game-engine/mechanics/powerup-spawner/index.ts]: This code is really close to the `handleFoodSpawning` maybe we can extract the common logic? (And also consolidate tests)

File: game-backend/src/game-engine/state-selectors/helper.ts [hunk 4]
```diff
@@ -17,6 +18,16 @@ const getNearestObjectOfType = <T extends GameObject>(
...
     .sort((a, b) => (a.distance < b.distance ? -1 : 1));
+  return sortedObjects;
+};
+
+const getNearestObjectOfType = <T extends GameObject>(
+  state: State,
+  object: GameObject,
+  type: string
+): T => {
+  const objects = getGameObjectsPerType(state, type) as T[];
```
Comment (StuckiSimon) [inline game-backend/src/game-engine/state-selectors/helper.ts:29]: Try not to type cast - this erases the benefit of TypeScript.

File: game-backend/src/helper/consumable.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+  State,
+} from '@darwin/types';
+import { powerupList } from '../game-engine/mechanics/powerup-spawner/powerups/index';
+import { getObjectById } from './gameObjects';
+
+export const isConsumable = (gameObject: GameObject): boolean =>
+  [GameObjectTypes.Food, ...powerupList].indexOf(gameObject.type) >= 0;
+
+export const filterConsumables = (gameObjects: GameObject[]): Consumable[] =>
+  gameObjects.filter(isConsumable) as Consumable[];
```
Comment (StuckiSimon) [inline game-backend/src/helper/consumable.ts]: Try not to type cast - this erases the benefit of TypeScript.

File: game-backend/src/helper/gameObjects.ts [hunk 6]
```diff
@@ -51,8 +51,10 @@ export const createUnit = ({
   type: GameObjectTypes.Unit,
 });
 
-const getObjectById = (state: State, id: ObjectId): GameObject =>
-  state.objectMap[id];
+export const getObjectById = <T extends GameObject>(
+  state: State,
+  id: ObjectId
+): T => state.objectMap[id] as T;
```
Comment (StuckiSimon) [inline game-backend/src/helper/gameObjects.ts:57]: Try not to type cast - this erases the benefit of TypeScript. Maybe adding an assertion here would make sense.

File: game-backend/src/helper/powerup.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+  GameObjectTypes,
+  Consume,
+  GameObject,
+} from '@darwin/types';
+import powerupMap from '../game-engine/mechanics/powerup-spawner/powerups/index';
+
+export const filterPowerups = (objects: GameObject[]): Consumable[] =>
+  objects.filter(
+    obj => !!(powerupMap as Record<GameObjectTypes, Consume>)[obj.type]
+  ) as Consumable[];
```
Comment (StuckiSimon) [inline game-backend/src/helper/powerup.ts]: Try not to type cast - this erases the benefit of TypeScript. This is type cast sorcery 😅

File: game-frontend/src/components/canvas-objects/Powerup.tsx [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+const radius = HALF_FIELD_SIZE * 0.3;
+
+// Color inspiration: https://coolors.co/337fe6-f46547-05bf96-fcb136-a07be6
+const COLOR_MAP: Record<PowerupType, number> = {
+  enduranceBoost: 0x05bf96,
+};
+
+const PowerUp: FC<Props> = ({ consumable }) => {
+  const color = COLOR_MAP[consumable.type as PowerupType];
```
Comment (StuckiSimon) [inline game-frontend/src/components/canvas-objects/Powerup.tsx]: What about steering this from outside using props instead of an inside lookup map? This would make this type casting unnecessary - and it would prevent users from forgetting to update the map.

File: game-backend/src/helper/array.ts [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+    [array[i], array[j]] = [array[j], array[i]];
+  }
+  return array;
+}
+
+export const pickRandom = <T>(array: Array<T>): T => {
+  const { length } = array;
+  const randomIndex = Math.floor(Math.random() * length);
+  return array[randomIndex];
+};
```
Comment (StuckiSimon) [inline game-backend/src/helper/array.ts:18]: Why not write tests for these kind of features?

**Review by StuckiSimon — COMMENTED** [review]
Check your typings, there are quite a lot of type casts in the code.
Also watch test coverage and code duplicates.

File: game-backend/src/game-engine/mechanics/createSpawner.ts [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+import { generateFreePosition } from '../../helper/fields';
+import produce from '../../helper/produce';
+
+type params = {
+  limit: (state: State) => number;
+  count: (state: State) => number;
+  spawn: (position: Position) => GameObject;
+};
+
+const createSpawner = ({ limit, count, spawn }: params) => (
```
Comment (StuckiSimon) [inline game-backend/src/game-engine/mechanics/createSpawner.ts:11]: Now that it's extracted, there should be a unit test for `createSpawner.ts` but currently it's tested twice using the separate tests. What's the benefit of testing it for each usage again?

**Review by StuckiSimon — COMMENTED** [review]
Please keep an eye on these tests, otherwise looks finally ready to merge 🎉

**Review by StuckiSimon — APPROVED** [review]
Looks good 👍
